main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1155 @@
+        exp1 <- visitExpression(exp10, kenv, root)
+        exp2 <- visitExpression(exp20, kenv, root)
+      } yield KindedAst.Expression.Binary(sop, exp1, exp2, Type.freshVar(Kind.Star), loc)
+
+    case ResolvedAst.Expression.IfThenElse(exp10, exp20, exp30, loc) =>
+      for {
+        exp1 <- visitExpression(exp10, kenv, root)
+        exp2 <- visitExpression(exp20, kenv, root)
+        exp3 <- visitExpression(exp30, kenv, root)
+      } yield KindedAst.Expression.IfThenElse(exp1, exp2, exp3, loc)
+
+    case ResolvedAst.Expression.Stm(exp10, exp20, loc) =>
+      for {
+        exp1 <- visitExpression(exp10, kenv, root)
+        exp2 <- visitExpression(exp20, kenv, root)
+      } yield KindedAst.Expression.Stm(exp1, exp2, loc)
+
+    case ResolvedAst.Expression.Let(sym, mod, exp10, exp20, loc) =>
+      for {
+        exp1 <- visitExpression(exp10, kenv, root)
+        exp2 <- visitExpression(exp20, kenv, root)
+      } yield KindedAst.Expression.Let(sym, mod, exp1, exp2, loc)
+
+    case ResolvedAst.Expression.LetRegion(sym, exp0, loc) =>
+      for {
+        exp <- visitExpression(exp0, kenv, root)
+      } yield KindedAst.Expression.LetRegion(sym, exp, Type.freshVar(Kind.Bool), loc)
+
+    case ResolvedAst.Expression.Match(exp0, rules0, loc) =>
+      for {
+        exp <- visitExpression(exp0, kenv, root)
+        rules <- traverse(rules0)(visitMatchRule(_, kenv, root))
+      } yield KindedAst.Expression.Match(exp, rules, loc)
+
+    case ResolvedAst.Expression.Choose(star, exps0, rules0, loc) =>
+      for {
+        exps <- Validation.traverse(exps0)(visitExpression(_, kenv, root))
+        rules <- Validation.traverse(rules0)(visitChoiceRule(_, kenv, root))
+      } yield KindedAst.Expression.Choose(star, exps, rules, Type.freshVar(Kind.Star), loc)
+
+    case ResolvedAst.Expression.Tag(sym, tag, exp0, loc) =>
+      for {
+        exp <- visitExpression(exp0, kenv, root)
+      } yield KindedAst.Expression.Tag(sym, tag, exp, Type.freshVar(Kind.Star), loc)
+
+    case ResolvedAst.Expression.Tuple(elms0, loc) =>
+      for {
+        elms <- Validation.traverse(elms0)(visitExpression(_, kenv, root))
+      } yield KindedAst.Expression.Tuple(elms, loc)
+
+    case ResolvedAst.Expression.RecordEmpty(loc) => KindedAst.Expression.RecordEmpty(Type.freshVar(Kind.Record), loc).toSuccess
+
+    case ResolvedAst.Expression.RecordSelect(exp0, field, loc) =>
+      for {
+        exp <- visitExpression(exp0, kenv, root)
+      } yield KindedAst.Expression.RecordSelect(exp, field, Type.freshVar(Kind.Star), loc)
+
+    case ResolvedAst.Expression.RecordExtend(field, value0, rest0, loc) =>
+      // Ideally, if `rest` is not of record kind, we should throw a kind error.
+      // But because we have subkinding, we can't do this in the Kinder.
+      // Consider: { +name = 5 | id({}) }
+      // This is OK, but would be seen as a kind error since id is `(a: *) -> (a: *)`, so `id({}) :: *`
+      // This KindError will be caught later in the Typer
```

Comment:
Do we have a path towards correct kinding for records/schemas? (This is completely swapped out of my mind.)

Reply:
The latest ideas on that front are here: https://github.com/flix/flix/issues/2057